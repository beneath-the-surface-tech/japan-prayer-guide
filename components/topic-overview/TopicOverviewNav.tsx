import { useTranslation } from "next-i18next"
import { Dispatch, SetStateAction, useEffect, useRef, useState } from "react"
import { Container } from "react-bootstrap"

interface Props {
  selected: string
  setSelected: Dispatch<SetStateAction<string>>
}

const scrollOffset = -50

export const TopicOverviewNav = ({ selected, setSelected }: Props) => {
  const { t } = useTranslation("topic-overview")

  const bannerRef = useRef(null)
  const [show, setShow] = useState(false)
  const isScrolling = useRef(false)

  const highlightCurrentNavLink = () => {
    // If we're not in the 'church' section, show 'culture' as selected by default
    const scrollY = window.scrollY

    const churchSection = document.getElementById("church")

    if (churchSection) {
      // The leeway makes it so that if they click the link and scroll a little, it will still show the correct section as selected
      const scrollLeeway = 10
      const churchPosition =
        churchSection.getBoundingClientRect().top + window.pageYOffset + scrollOffset - scrollLeeway

      // Determine which section is active based on the scroll position
      if (scrollY >= churchPosition) {
        setSelected("church")
      } else {
        setSelected("culture")
      }
    }
  }

  useEffect(() => {
    const handleScroll = () => {
      const bannerElement: any = bannerRef.current
      if (bannerElement) {
        const { top } = bannerElement.getBoundingClientRect() + window.pageYOffset
        if (top === 0) {
          setShow(true)
        } else {
          setShow(false)
        }
      }

      if (!isScrolling.current) {
        highlightCurrentNavLink()
      }
    }

    window.addEventListener("scroll", handleScroll)
    return () => {
      window.removeEventListener("scroll", handleScroll)
    }
  }, [])

  const scrollToCustomPosition = (id: string) => {
    setSelected(id)
    isScrolling.current = true // block the listener from triggering while we scroll
    const targetElement = document.getElementById(id)

    if (targetElement) {
      const topPosition = targetElement.getBoundingClientRect().top
      window.scrollTo({ top: topPosition + window.pageYOffset + scrollOffset, behavior: "smooth" })

      setTimeout(() => {
        isScrolling.current = false
      }, 1500)
    }
  }

  return (
    <>
      <Container
        style={{ maxWidth: "none" }}
        ref={bannerRef}
        data-testid={"overview-nav-container"}
        className="bottom-grey-border w-100 d-flex flex-column align-items-center justify-content-center sticky-top p-0"
      >
        <div id="overview-nav" className="w-100 opacity-0" style={show ? {} : { height: "0px" }}></div>
        <div
          data-testid={"overview-nav-links"}
          className="pt-4 pb-0 w-100 d-flex justify-content-center gap-3 bg-white"
        >
          <a
            className={
              "topic-nav-tab px-3 my-0 text-decoration-none" +
              (selected === "culture"
                ? " border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
                : " text-black")
            }
            onClick={() => scrollToCustomPosition("culture")}
            data-testid={"overview-nav-link"}
          >
            {t("topHeading")}
          </a>
          <a
            className={
              "topic-nav-tab px-3 my-0 text-decoration-none " +
              (selected === "church"
                ? " border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
                : " text-black")
            }
            onClick={() => scrollToCustomPosition("church")}
            data-testid="overview-nav-link-2"
          >
            {t("botHeading")}
          </a>
        </div>
      </Container>
    </>
  )
}
