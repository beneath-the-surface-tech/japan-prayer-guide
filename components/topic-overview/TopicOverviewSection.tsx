import Link from "next/link"
import { Card, Container, Carousel, Row, Col, Image } from "react-bootstrap";

interface TopicOverviewProps {
    title: string
    links: string[]
    section: string
    labels: string[]
}

export const TopicOverviewSection = ({title, links, labels, section}:TopicOverviewProps) => {

    return (
        <>
            <section>
                <Container id={section} className="no-max-container mt-4 mt-md-5">
                    <h1 className="topic-nav-section-title text-primary pt-4 px-0 px-md-2 mb-0">{title}</h1>
                </Container>
                <Container className="bottom-grey-border px-4 pt-1 no-max-container pb-4">
                    <Row xl={3} lg={3} md={2} sm={2} xs={2}>
                        {labels.map((element, idx) => {
                            // Special if statement for conference, the 'else' will be removed for final
                            if ((idx === 0 && section === "culture") || (idx === 2 && section === "church")) {
                                return (
                                    <Col key={element + idx} className="d-flex justify-content-center">
                                        <Link href={`/topics/${links[idx].toLowerCase()}`} className="text-decoration-none">
                                            <Card className="topic-nav-card mx-1 my-3">
                                                <Image
                                                    src={`/photos/topic-nav/${section}/${links[idx]}.png`}
                                                    alt={""}
                                                />
                                                <Card.Body style={{height: 'fit-content'}} className="d-flex align-items-center topic-nav-card-title">
                                                    <p className="m-0">{element}</p>
                                                </Card.Body>
                                            </Card>
                                        </Link>
                                    </Col>
                                )
                            } else {
                                return (
                                    <Col key={element + idx} className="d-flex justify-content-center">
                                        <Link href={''} className="text-decoration-none position-relative">
                                            <Card className="topic-nav-card mx-1 my-3">
                                                <div className="position-relative" style={{width: 'fit-content', height: 'fit-content'}}>
                                                    <Image
                                                        src={`/photos/topic-nav/${section}/placeholder.png`}
                                                        alt={""}
                                                        style={{aspectRatio: 1.772, objectFit: 'cover', width: '100%'}}
                                                    />
                                                    <div style={{zIndex: 2, top: 0, cursor: 'not-allowed', backgroundColor: '#727C96'}} className="position-absolute w-100 h-100 opacity-75"></div>
                                                </div>
                                                <Card.Body className="d-flex align-items-center topic-nav-card-title" style={{cursor: 'not-allowed'}}>
                                                    <p className="m-0">{element}</p>
                                                </Card.Body>
                                            </Card>
                                        </Link>
                                    </Col>
                                )
                            }
                        })}
                    </Row>
                </Container>
            </section>
        </>
    )
}