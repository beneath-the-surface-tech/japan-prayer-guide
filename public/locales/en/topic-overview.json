{
  "webpageTitle": "30 ways to pray for Japan",
  "heading": "Prayer topics",
  "pageSubtitle": [
    "After surveying missionaries, pastors, and Christian leaders in Japan, we collected over 500 prayer points and grouped them into topics for Japanese culture and society, and church and mission.",
    "The 30 topics chosen were those prioritized by Japanese leaders as key opportunities and challenges for the gospel in Japan today."
  ],
  "topHeading": "Culture and society",
  "botHeading": "Church and mission",
  "cultureLinks": [
    "popular-religious-practices",
    "japanese-views-of-chrisianity",
    "japansese-culture-and-christianity",
    "ancestor-veneration",
    "harmony",
    "social-conformity",
    "self-reliance-and-materialism",
    "mental-and-emotional-health",
    "aging-society",
    "politics-and-international-relations",
    "natural-disasters",
    "technology-and-media",
    "workplace-pressure",
    "family-life",
    "youth-and-young-adults"
  ],
  "cultureTopics": [
    "Popular religious practices",
    "Japanese views of Chrisianity",
    "Japansese culture and Christianity",
    "Ancestor veneration",
    "Harmony",
    "Social conformity",
    "Self reliance and materialism",
    "Mental and emotional health",
    "Aging society",
    "Politics and international relations",
    "Natural disasters",
    "Technology and media",
    "Workplace pressure",
    "Family life",
    "Youth and young adults"
  ],
  "churchLinks": [
    "struggling-churches",
    "church-planting",
    "church-leadership",
    "church-in-the-community",
    "creative-ways-of-ministry",
    "christianity-in-rural-areas",
    "personal-evangelism",
    "faith-in-the-workplace",
    "daily-christian-life-and-faith",
    "japanese-overseas",
    "missionary-life",
    "unity-in-christ",
    "japanese-mission-movement",
    "prayer-for-spiritual-breakthrough"
  ],
  "churchTopics": [
    "Struggling churches",
    "Church planting",
    "Church leadership",
    "Church in the community",
    "Creative ways of ministry",
    "Christianity in rural areas",
    "Personal evangelism",
    "Faith in the workplace",
    "Daily Christian life and faith",
    "Japanese overseas",
    "Missionary life",
    "Unity in Christ",
    "Japanese mission movement",
    "Prayer for spiritual breakthrough"
  ]
}
