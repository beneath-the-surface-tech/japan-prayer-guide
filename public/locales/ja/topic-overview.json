{
    "webpageTitle": "日本のために祈る３０項目",
    "heading": "祈りのテーマ",
    "pageSubtitle": [
        "日本在住の宣教師、牧師、クリスチャンリーダーたちにアンケートを取ったところ、500以上もの祈祷課題が集まりました。「御心が地にもなるように」では、それらを日本の文化や社会、そして教会や宣教に基づいてテーマごとに分類しています。",
        "選ばれた30のテーマは、今日の日本における福音の重要な機会および課題として、日本のクリスチャンリーダーたちが優先的に取り上げたものです。",
        "",
        "準備中。\n2024年にウェブサイト完全公開予定。"
    ],
    "cultureHeading": "文化と社会",
    "churchHeading": "教会と宣教",
    "cultureLinks": [
        "popular-religious-practices",
        "japanese-views-of-chrisianity",
        "japansese-culture-and-christianity",
        "ancestor-veneration",
        "harmony",
        "social-conformity",
        "self-reliance-and-materialism",
        "mental-and-emotional-health",
        "aging-society",
        "politics-and-international-relations",
        "immigration-and-foreign-nationals",
        "natural-disasters",
        "technology-and-media",
        "workplace-pressure",
        "family-life",
        "youth-and-young-adults"
    ],
    "cultureTopics": [
        "日本人の宗教的習慣",
        "日本人のキリスト教観",
        "日本文化とキリスト教",
        "先祖崇拝",
        "和を重んじる文化",
        "社会の同調圧力",
        "自分に頼る生き方と物質主義",
        "精神と心の健康",
        "高齢化社会",
        "政治と国際関係",
        "入国管理と外国人居住者",
        "自然災害",
        "テクノロジーとメディア",
        "仕事のプレッシャー",
        "家庭生活",
        "若者たち"
    ],
    "churchLinks": [
        "struggling-churches",
        "church-planting",
        "church-leadership",
        "church-in-the-community",
        "creative-ways-of-ministry",
        "christianity-in-rural-areas",
        "personal-evangelism",
        "faith-in-the-workplace",
        "daily-christian-life-and-faith",
        "japanese-overseas",
        "missionary-life",
        "unity-in-christ",
        "japanese-mission-movement",
        "prayer-for-spiritual-breakthrough"
    ],
    "churchTopics": [
        "伸び悩む教会",
        "教会開拓",
        "教会のリーダーシップ",
        "地域社会における教会",
        "創造的なミニストリー",
        "地方伝道",
        "個人伝道",
        "職場における信仰生活",
        "日常生活と信仰",
        "海外に住む日本人",
        "宣教師の生活",
        "キリストにおける一致",
        "世界宣教への関わり",
        "霊的突破のための祈り"
    ]
}
