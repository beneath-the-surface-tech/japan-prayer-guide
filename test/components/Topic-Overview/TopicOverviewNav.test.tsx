import { render, screen } from "@testing-library/react"
import { TopicOverviewNav } from "../../../components/topic-overview/TopicOverviewNav"
import React from "react"

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}))
jest.mock("react-i18next", () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str) => str,
      i18n: {
        changeLanguage: () => new Promise(() => {}),
      },
    }
  },
  initReactI18next: {
    type: "3rdParty",
    init: () => {},
  },
}))

describe("Overview Nav", () => {
  test("Renders a section with the right text inside", () => {
    const setSelected = jest.fn()
    const selected = "church"
    render(<TopicOverviewNav selected={selected} setSelected={setSelected} />)
    const navContainer = screen.getByTestId("overview-nav-container")

    expect(navContainer).toHaveClass("sticky-top")
  })
})
