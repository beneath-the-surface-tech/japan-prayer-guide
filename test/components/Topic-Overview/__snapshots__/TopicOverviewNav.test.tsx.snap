// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview Nav Snapshot test 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bottom-grey-border w-100 d-flex flex-column align-items-center justify-content-center sticky-top p-0 container"
        data-testid="overview-nav-container"
        style="max-width: none;"
      >
        <div
          class="w-100 opacity-0"
          id="overview-nav"
          style="height: 0px;"
        />
        <div
          class="pt-4 pb-0 w-100 d-flex justify-content-center gap-3 bg-white"
          data-testid="overview-nav-links"
        >
          <a
            class="common-p topic-nav-tab px-3 my-0 text-decoration-none border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
            data-testid="overview-nav-link"
          >
            topHeading
          </a>
          <a
            class="common-p topic-nav-tab px-3 my-0 text-decoration-none  text-black"
            data-testid="overview-nav-link-2"
          >
            botHeading
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bottom-grey-border w-100 d-flex flex-column align-items-center justify-content-center sticky-top p-0 container"
      data-testid="overview-nav-container"
      style="max-width: none;"
    >
      <div
        class="w-100 opacity-0"
        id="overview-nav"
        style="height: 0px;"
      />
      <div
        class="pt-4 pb-0 w-100 d-flex justify-content-center gap-3 bg-white"
        data-testid="overview-nav-links"
      >
        <a
          class="common-p topic-nav-tab px-3 my-0 text-decoration-none border-bottom border-secondary-5 border-3 text-secondary-5 fw-bold"
          data-testid="overview-nav-link"
        >
          topHeading
        </a>
        <a
          class="common-p topic-nav-tab px-3 my-0 text-decoration-none  text-black"
          data-testid="overview-nav-link-2"
        >
          botHeading
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
